move apply from apply to aggregate

thewilsonator opened the pull request:
apply is only used by aggregate.d and dclass.d and doesn't really belong in apply which should probably be named postorderVisitor

thewilsonator commented:
urgh
```
src\dmd\tocvdebug.d(561): Error: function `dmd.aggregate.apply(Dsymbol symbol, int function(Dsymbol s, AggregateDeclaration agg) fp, AggregateDeclaration agg)` is not callable using argument types `(Dsymbol, extern (C++) int function(Dsymbol s, CvFieldList* pmc), CvFieldList*)`
src\dmd\tocvdebug.d(561):        cannot pass argument `& cv_mem_count` of type `extern (C++) int function(Dsymbol s, CvFieldList* pmc)` to parameter `int function(Dsymbol s, AggregateDeclaration agg) fp`
src\dmd\tocvdebug.d(572): Error: function `dmd.aggregate.apply(Dsymbol symbol, int function(Dsymbol s, AggregateDeclaration agg) fp, AggregateDeclaration agg)` is not callable using argument types `(Dsymbol, extern (C++) int function(Dsymbol s, CvFieldList* pmc), CvFieldList*)`
src\dmd\tocvdebug.d(572):        cannot pass argument `& cv_mem_p` of type `extern (C++) int function(Dsymbol s, CvFieldList* pmc)` to parameter `int function(Dsymbol s, AggregateDeclaration agg) fp`
src\dmd\tocvdebug.d(724): Error: function `dmd.aggregate.apply(Dsymbol symbol, int function(Dsymbol s, AggregateDeclaration agg) fp, AggregateDeclaration agg)` is not callable using argument types `(Dsymbol, extern (C++) int function(Dsymbol s, CvFieldList* pmc), CvFieldList*)`
src\dmd\tocvdebug.d(724):        cannot pass argument `& cv_mem_count` of type `extern (C++) int function(Dsymbol s, CvFieldList* pmc)` to parameter `int function(Dsymbol s, AggregateDeclaration agg) fp`
src\dmd\tocvdebug.d(775): Error: function `dmd.aggregate.apply(Dsymbol symbol, int function(Dsymbol s, AggregateDeclaration agg) fp, AggregateDeclaration agg)` is not callable using argument types `(Dsymbol, extern (C++) int function(Dsymbol s, CvFieldList* pmc), CvFieldList*)`
src\dmd\tocvdebug.d(775):        cannot pass argument `& cv_mem_p` of type `extern (C++) int function(Dsymbol s, CvFieldList* pmc)` to parameter `int function(Dsymbol s, AggregateDeclaration agg) fp`
```

WalterBright commented on src/dmd/aggregate.d:
```diff
@@ -866,4 +863,38 @@ extern (C++) abstract class AggregateDeclaration : ScopeDsymbol
+ *  symbol = the dsymbol or parent of members to call fp on
+ *  fp = function pointer to process the iterated symbol.
+ *       If it returns nonzero, the iteration will be aborted.
+ *  params = any parameters passed to fp.
```
it's been changed to `agg`

thewilsonator replied:
I'll be changing that back soon

thewilsonator committed:
move apply from apply to aggregate

apply is only used by aggregate.d, dclass.d, and tocvdebug.d and doesn't really belong in apply which should probably be named postorderVisitor

thewilsonator commented:
@RazvanN7 pong. Apologies for the late response.

RazvanN7 commented on src/dmd/aggregate.d:
```diff
@@ -814,3 +811,36 @@ extern (C++) abstract class AggregateDeclaration : ScopeDsymbol
+ *  nonzero if the iteration is aborted by the return value of fp,
+ *  or 0 if it's completed.
+ */
+int apply(Dsymbol symbol, int function(Dsymbol, void*) fp, void* ctx)
```
Why not use the variadic parameter? It looks nicer than having `void*` in the signature. Are you worried about the size of the binary? It seems like there are 3 different instantiations only.

RazvanN7 replied:
@thewilsonator  ?

RazvanN7 replied:
Still, I think it's better to use the variadic. But I am willing to let this slide.

thewilsonator commented:
Thanks!
